### ISSUE TITLE
Product data validation fails when trying to import the products to update existing products data.

### ISSUE THREAD

**Issue by Taral-Patoliya** [firegento/FireGento_FastSimpleImport2#18]
When i use this module to import the products, it imports the products correctly, but when i use the same data again, instead of updating the product data, it generates validation error having all the rows invalid.

This happens only when i use this modul, if i generate a CSV file and import it manually using core module, it imports the products correctly, and updates the products.

**Comment by EliasKotlyar** [comment]
Hello Taral-Patoliya,

Could you please provide a simple Script in which this Error occurs? I could not reproduce it with the demo-module.

Greetings
Elias

**Comment by Taral-Patoliya** [comment]
@EliasKotlyar , i have traced and solved this issue, it was a magento 2 core issue.
The script where the url keys are inserted are not checked if the same key already exists. i changed a little bit and the import started to work flawlessly.

**Comment by EliasKotlyar** [comment]
Hello Taral-Patoliya,

Many Thanks for the explanation. Could you please provide this Fix? Did you pushed it to the M2 Repository?

Greetings
Elias

**Comment by Taral-Patoliya** [comment]
@EliasKotlyar  I have not pushed it to the M2 repository.
Please check the fix in the following code snippet. 
In the class  `Magento\CatalogUrlRewrite\Observer\AfterImportDataObserver`
 
I just changed the execute method of the observer which generates the url keys and inserts them to the database.
```
public function execute(Observer $observer)
    {

        $this->import = $observer->getEvent()->getAdapter();
        if ($products = $observer->getEvent()->getBunch()) {
            foreach ($products as $product) {
                $this->_populateForUrlGeneration($product);
            }
            $temp = $productUrls = $this->generateUrls();

            foreach($temp as $key => $url)
            {
                $array = [
                    'entity_type'=>$url->getEntityType(),
                    'entity_id'=>$url->getEntityId(),
                    'request_path'=>$url->getRequestPath(),
                    'target_path'=>$url->getTargetPath(),
                    'store_id'=>$url->getStoreId()
                    /*'metadata'=>$url->getMetadata()*/
                ];
                if($this->checkIfUrlExists($array)){
                	unset($productUrls[$key]);
                }
            }
            if ($productUrls) {
                $this->urlPersist->replace($productUrls);
            }
        }
    }

 protected function checkIfUrlExists($data)
    {
    	return (bool)$this->connection->fetchRow($this->prepareSelect($data));
    }

    /**
     * Prepare select statement for specific filter
     *
     * @param array $data
     * @return \Magento\Framework\DB\Select
     */
    protected function prepareSelect($data)
    {
        $select = $this->connection->select('url_rewrite_id');
        $select->from($this->resource->getTableName('url_rewrite'));

        foreach ($data as $column => $value) {
            $select->where($this->connection->quoteIdentifier($column) . ' = ?', $value);
        }
        return $select;
    }
```

I am checking each url by fetching one record from the database based on the contents of the generated url key object and if it is found in the database, then it is skipped from the array of the url keys to be inserted. This is still a half baked solution, i am open to new suggestions.

Thank you.

**Comment by rhm13** [comment]
@Taral-Patoliya, I'm having this exact issue when attempting to import 2,000 products via a csv file:

>3: Specified URL key already exists in row(s): 2794, 2795, 2796, 2797, 2798, 2799, 2800, 2801, 2802, >2803, 2804, 2805, 2806, 2807, 2808, 2809, 28                           10, 2811, 2812, 2813, 2814, 2815, 2816, >2817, 2818, 2819, 2820, 2821, 2822, 2823, 2824, 2825, 2826, 2827, 2828, 2829, 2830, 2831, 2832, 2833, >283                           4, 2835, 2836, 2837, 2838, 2839, 2840, 2841, 3712, 3713, 3714, 3715, 3716, 3717, >3718, 3719, 3720, 3721, 3722, 3723, 3724, 3725, 3726, 3727, 3728                           , 3729, 3730, 3731, >3732, 3733, 3734, 3735, 3736, 3737, 3738, 3739, 3740, 3741, 3742, 3743, 3744, 3745, 3746, 3747, 3748, >3749, 3750, 3751, 3752,                            3753, 3754, 3755, 3756, 3757, 3758, 3759, 3760, 3761, 3762, >3763, 3764, 3765, 3766, 3767, 3768, 3769, 3770, 3771, 3772, 3773, 3774, 3775, 3776,                            >3777, 3778, 3779, 3780, 3781, 3782, 3783, 3784, 3785, 3786, 3787, 3788, 3789, 3790, 3791, 3792, 3793, >3794, 3795, 3796, 3797, 3798, 3799, 3800, 3                           801, 3802, 3803, 3804, 3805, 3806, 3807, >3808, 3809, 3810, 3811, 3812, 3813, 3814, 3815, 3816, 3817, 3818, 3819, 3820, 3821, 3822, 3823, 3824, >38                           25, 3826, 3827, 3828, 3829, 3830, 3831, 3832, 3833, 3834, 3835, 3836, 3837, 3838, >3839, 3840, 3841, 3842, 3843, 3844, 3845, 3846, 8701, 8702, 870                           3, 8704, 8705, 8706, >8707, 8708, 8709, 17350, 17351, 17352, 17365, 17366, 17367, 22732, 22733, 22734, 22738, 22739, >22740, 22750, 22751, 22752, 2                           2756, 22757, 22758, 22768, 22769, 22770, 22771, 22772, >22773, 22786, 22787, 22788, 22789, 22790, 22791, 23590, 23591, 23592, 23596, 23597, >23598,                            23599, 23600, 23601, 28735, 28736, 28737, 28738, 28739, 28740, 28744, 28745, >28746
>4: Attribute with this code is not super in row(s): 33295
>5: Checked rows: 33295, checked entities: 33295, invalid rows: 241, total errors: 241

I tried inserting your code that pushes the array of the url keys to be inserted, but I still got the same error. I've sanitized the data as much as is humanly possible in the csv, so if you make more head-way with this issue I'd be massively grateful! 👍 Many thanks.

**Comment by ericrisler** [comment]
@ Taral-Patoliya  - just a note that we encountered the same and added a pre-processing layer to check ensure that url-keys are unique'd before passing to the magento importer. Your solution may be better, but I believe there are issues open to the M2 developers already and they may come up with something similar.

**Comment by Taral-Patoliya** [comment]
Yes, You are right @ericrisler, this is an open issue with magento and not solved in the latest version of CE.

**Comment by dzschille** [comment]
@Taral-Patoliya do you mean this [magento issue](https://github.com/magento/magento2/issues/8396)? This issue is already fixed in 2.2.0. Has anybody still this error?
